// coinpayments/dist/validation/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterMassWithdrawalOpts = exports.validateMassWithdrawalOpts = exports.hasOneOfFields = exports.validatePayload = exports.validateMassWithDrawal = exports.validate = exports.getValidationSchema = exports.getVariations = void 0;
var error_1 = __importDefault(require("../error"));
var schema_1 = __importDefault(require("./schema"));
var constants_1 = require("../constants");
var getVariations = function (validationSchema) {
    var variations = validationSchema.find(function (key) { return Array.isArray(key); });
    var staticKeys = validationSchema.filter(function (key) { return !Array.isArray(key); });
    if (!(variations === null || variations === void 0 ? void 0 : variations.length)) {
        return [validationSchema];
    }
    return variations.map(function (variation) {
        return [].concat(staticKeys).concat([variation]);
    });
};
exports.getVariations = getVariations;
var getValidationSchema = function (cmd) {
    if (!cmd) {
        throw new error_1.default('No method passed');
    }
    var validationSchema = schema_1.default[cmd];
    if (!validationSchema) {
        throw new error_1.default("No such method ".concat(cmd));
    }
    return validationSchema;
};
exports.getValidationSchema = getValidationSchema;
var validate = function (validationSchema, options) {
    var optionKeys = Object.keys(options);
    var validationRules = (0, exports.getVariations)(validationSchema);
    return validationRules.some(function (rule) { return rule.every(function (r) { return optionKeys.includes(r); }); });
};
exports.validate = validate;
var validateMassWithDrawal = function (options) {
    var regex = /(wd\[wd[0-9]*\]\[(amount|address|currency|dest_tag|domain|pbntag|add_tx_fee|currency2|ipn_url|auto_confirm|note)\]|cmd)/;
    return Object.keys(options).every(function (key) { return regex.test(key); });
};
exports.validateMassWithDrawal = validateMassWithDrawal;
var validatePayload = function (options) {
    if (options.cmd === constants_1.CMDS.CREATE_MASS_WITHDRAWAL) {
        if (!(0, exports.validateMassWithDrawal)(options)) {
            throw new error_1.default('Invalid mass withdrawal payload', options);
        }
    }
    var validationSchema = (0, exports.getValidationSchema)(options.cmd);
    if (!(0, exports.validate)(validationSchema, options)) {
        throw new error_1.default('Missing parameters', options);
    }
};
exports.validatePayload = validatePayload;
var hasOneOfFields = function (obj, keys) { return keys.filter(function (key) { return obj.hasOwnProperty(key); }).length === 1; };
exports.hasOneOfFields = hasOneOfFields;
var validateMassWithdrawalOpts = function (withdrawalOpts) {
    return withdrawalOpts.currency &&
        withdrawalOpts.amount &&
        (0, exports.hasOneOfFields)(withdrawalOpts, ['address', 'domain', 'pbntag']);
};
exports.validateMassWithdrawalOpts = validateMassWithdrawalOpts;
var filterMassWithdrawalOpts = function (massWithdrawalOpts) {
    return massWithdrawalOpts.filter(exports.validateMassWithdrawalOpts);
};
exports.filterMassWithdrawalOpts = filterMassWithdrawalOpts;
